### DISCUSSION TITLE
Is it possible to check multiple conditions for conditional rendering?

### DISCUSSION THREAD

**Discussion by SamEdwardes** [quarto-dev/quarto-cli#9295] [Q&A] [answered]
### Description

Is it possible to check multiple conditions for conditional rendering? For example, can I render some sections for HTML or ipynb, while another section for PDF or docx?

``````md
---
title: Reprex
execute:
  echo: false
format:
  html:
    code-tools: true
  pdf: default
  docx: default
  ipynb:
    echo: true
---

This is always visible.

::: {.content-visible when-format="html|ipynb"}
```{python}
print("This is in HTML or ipynb")
```
:::

::: {.content-visible when-format="pdf|docx"}
```{python}
print("This is in PDF or docx")
```
:::
``````

**Comment by mcanouil** [comment]
Don't combine the values, add any number of when/unless statements.

**Reply by cderv** [reply]
@mcanouil I don't think this will work to use several `when-format`. Did you try ? 

Putting several attributes may be currently handled for this feature and only last will apply . 

Meaning 
````
::: {.content-visible when-format="pdf" when-format="docx"}
```{python}
print("This is in PDF or docx")
```
:::

````

Only shows in `docx`

I think we need to improve this or come up with a syntax for multiple format.

**Reply by mcanouil** [reply]
Indeed, I was pretty convinced I did that in the past (not only combining when/unless).

**Reply by cscheid** [reply]
This is something we could consider improving, but I don't know how to make the semantics easy to explain when repeating precisely the same key (`when-format`, etc), especially when combined with other keys.

**Reply by SamEdwardes** [reply]
Apologies I missed your replied earlier. Thank you for taking a look! I want to double check if I understand correctly. As of today, there is no way to express and/or logic to cover multiple formats. If I want to have more complex logic, my best bet would be to wrap up the logic in a function, and then have multiple divs to cover the rendering per format. For example...


``````markdown
---
title: Reprex
execute:
  echo: false
format:
  html:
    embed-resources: true
  pdf: default
  docx: default
  ipynb:
    echo: true
---


```{python}
import numpy as np
import matplotlib.pyplot as plt
```

```{python}
#| echo: false
#| output: false
def create_plot(title):
    r = np.arange(0, 2, 0.01)
    theta = 2 * np.pi * r
    fig, ax = plt.subplots(
        subplot_kw = {'projection': 'polar'}
    )
    ax.plot(theta, r)
    ax.set_rticks([0.5, 1, 1.5, 2])
    ax.grid(True)
    ax.set_title(title)
    plt.show()
```


::: {.content-visible when-format="html"}

Here is an HTML plot...

```{python}
create_plot('HTML rendered plot')
```

:::

::: {.content-visible when-format="ipynb"}

Here is an ipynb plot...

```{python}
create_plot('ipynb rendered plot')
```

:::

::: {.content-visible when-format="pdf"}

This is a PDF plot...

```{python}
create_plot('ipynb rendered plot')
```

:::

::: {.content-visible when-format="docx"}

This is a Docx plot...

```{python}
create_plot('Docx plot')
```

:::
``````

**Reply by SamEdwardes** [reply]
In the future it would be nice to express more complex logic. I do not know what the correct syntax would be, but something like this:

``````markdown
::: {.content-visible when-format="html" when-format="pdf"}

```{python}
create_plot('HTML and PDF rendered plot')
```

:::
``````

or

``````markdown
::: {.content-visible when-format="html|pdf"}

```{python}
create_plot('HTML and PDF rendered plot')
```

:::
``````

**Reply by cderv** [reply]
I believe we will solve this soon in some way. Idea is to pass the information of the output format name to engine using an environment variable, so that your python code could check for this and do accordingly. Your `create_plot()` function could then so different type of output depending on this environment variable. This would require the cell to be executed in quarto render context (which is not the default), but this would already be better. 

Do you think it would help if you were able to retrieve something like `QUARTO_PANDOC_TO` env var in your python code in cell ? 
`QUARTO_PANDOC_TO='docx'` or `QUARTO_PANDOC_TO='latex'` or `QUARTO_PANDOC_TO='html'` for examples

**Reply by SamEdwardes** [reply]
I think this is an interesting idea. It would handle some uses cases, for example create a dynamic plot when rendering to HTML, a static plot when rendering to PDF.

``````markdown

My plot:

```{python}
import os

def make_plot():
    if os.environ['QUARTO_PANDOC_TO'] == 'html':
        # make a plotly plot
        ...
    elif os.environ['QUARTO_PANDOC_TO'] == 'typst':
        # make a matplotlib plot
        ...
```
``````

I don't know if this would help with the conditional rendering of markdown though. For example, what if there is a certain div that I want to only render for word and pdf, but not for HTML?

**Reply by cderv** [reply]
Yes indeed, this is something that would need to be improved , but that has some challenged as commented above https://github.com/quarto-dev/quarto-cli/discussions/9295#discussioncomment-9050356

**Reply by mcanouil** [reply]
It's now possible to add several time the attribute:

- https://github.com/quarto-dev/quarto-cli/issues/11303

**Comment by SamEdwardes** [comment, marked as answer]
from @mcanouil:

> It's now possible to add several time the attribute:
>
> https://github.com/quarto-dev/quarto-cli/issues/11303

This should now work:

```quarto
::: {.content-hidden when-format="html" when-format="pdf"}

Should not appear in HTML or PDF.

:::
```
